runtime/major_gc.c

```diff
@@ -504,104 +503,149 @@ static void realloc_mark_stack (struct mark_stack* stk)
+                chd = Hd_val(child);
+                goto again;
+              }
+      } else {
+        atomic_store_explicit(
+          Hp_atomic_val(child),
+          With_status_hd(chd, global.MARKED),
+          memory_order_relaxed);
+      }
+      if(Tag_hd(chd) < No_scan_tag){
+        mark_stack_push(stk, child, 0, work);
+      }
+      else{
+        *work -= Wosize_hd(chd); /* account for header */
+      }
+    }
+  }
+  }
+}
+
+static intnat do_some_marking(intnat budget) {
+  struct mark_stack* stk = Caml_state->mark_stack;
   while (stk->count > 0) {
-    mark_entry e = stk->stack[--stk->count];
-    while (e.offset != e.end) {
-      value v;
+    mark_entry me = stk->stack[--stk->count];
+    intnat me_end = Wosize_val(me.block);
+    while (me.offset != me_end) {
       if (budget <= 0) {
-        budget -= mark_stack_push(stk, e);
+        mark_stack_push(stk, me.block, me.offset, NULL);
         return budget;
       }
       budget--;
-      CAMLassert(Is_markable(e.block) &&
-                 Has_status_hd(Hd_val(e.block), global.MARKED) &&
-                 Tag_val(e.block) < No_scan_tag &&
-                 Tag_val(e.block) != Cont_tag);
-      v = Op_val(e.block)[e.offset++];
-      if (Is_markable(v)) {
-        header_t hd = Hd_val(v);
-        if (Tag_hd(hd) == Infix_tag) {
-          v -= Infix_offset_hd(hd);
-          hd = Hd_val(v);
-        }
-        CAMLassert (!Has_status_hd(hd, global.GARBAGE));
-        if (Has_status_hd(hd, global.UNMARKED)) {
-          Caml_state->stat_blocks_marked++;
-          if (Tag_hd(hd) == Cont_tag) {
-            budget -= mark_stack_push(stk, e);
-            caml_darken_cont(v);
-            e = (mark_entry){0};
-            budget -= Wosize_hd(hd); /* credit for header, done with mark_entry */
-          } else {
-again:
-            if (Tag_hd(hd) == Lazy_tag || Tag_hd(hd) == Forcing_tag) {
-              if (!atomic_compare_exchange_strong(
-                    Hp_atomic_val(v), &hd,
-                    With_status_hd(hd, global.MARKED))) {
-                hd = Hd_val(v);
-                goto again;
-              }
-            }
-            else {
-              atomic_store_explicit(
-                Hp_atomic_val(v),
-                With_status_hd(hd, global.MARKED),
-                memory_order_relaxed);
-            }
-            if (Tag_hd(hd) < No_scan_tag) {
-              mark_entry child = {v, 0, Wosize_hd(hd)};
-              budget -= mark_stack_push(stk, e);
-              e = child;
-            } else {
-              budget -= Whsize_hd(hd);
-            }
-          }
-        }
+      CAMLassert(Is_markable(me.block) &&
+                 Has_status_hd(Hd_val(me.block), global.MARKED) &&
+                 Tag_val(me.block) < No_scan_tag &&
+                 Tag_val(me.block) != Cont_tag);
+      int is_cont = 0;
+      mark_slice_darken(stk, me.block, me.offset++, &budget, &is_cont);
+      if(is_cont){
```

Comment:
I think there is something not right about this. If we have a block with a continuation then we stop marking the remaining fields in the block. 

The testcase in ocaml-multicore#430 segfaults with this branch and I think the issue might be here.

Reply:
Fixed now, the testcase in #430 passes now on my machine. Perhaps we can kick off another CI run after #430 is merged.